File: app/src/main/java/se/gustavkarlsson/skylight/android/gui/activities/settings/SettingsFragment.java
package se.gustavkarlsson.skylight.android.gui.activities.settings;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceFragment;
import android.util.Log;

import se.gustavkarlsson.skylight.android.R;
import se.gustavkarlsson.skylight.android.Skylight;
import se.gustavkarlsson.skylight.android.background.UpdateScheduler;

public class SettingsFragment extends PreferenceFragment {
	private static final String TAG = SettingsFragment.class.getSimpleName();

	private NotificationsChangedListener notificationsChangedListener;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		Log.v(TAG, "onCreate");
		super.onCreate(savedInstanceState);

		UpdateScheduler updateScheduler = Skylight.getApplicationComponent(getActivity()).getUpdateScheduler();
		notificationsChangedListener = new NotificationsChangedListener(updateScheduler, getResources().getString(R.string.pref_notifications_key));
		getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener(notificationsChangedListener);
		addPreferencesFromResource(R.xml.preferences);
	}

	@Override
	public void onDestroy() {
		Log.v(TAG, "onDestroy");
		getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(notificationsChangedListener);
		notificationsChangedListener = null;
		super.onDestroy();
	}

	private static class NotificationsChangedListener implements SharedPreferences.OnSharedPreferenceChangeListener {
		private final UpdateScheduler updateScheduler;
		private final String notificationsKey;

		private NotificationsChangedListener(UpdateScheduler updateScheduler, String notificationsKey) {
			this.updateScheduler = updateScheduler;
			this.notificationsKey = notificationsKey;
		}

		@Override
		public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
			if (key.equals(this.notificationsKey)) {
				updateScheduler.setupBackgroundUpdates();
			}
		}
	}
}

File: app/src/main/java/se/gustavkarlsson/skylight/android/gui/activities/settings/SettingsFragment.kt
package se.gustavkarlsson.skylight.android.gui.activities.settings

import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceFragment
import android.util.Log

import se.gustavkarlsson.skylight.android.R
import se.gustavkarlsson.skylight.android.Skylight
import se.gustavkarlsson.skylight.android.background.UpdateScheduler

class SettingsFragment : PreferenceFragment() {

    private lateinit var notificationsChangedListener: NotificationsChangedListener

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.v(TAG, "onCreate")
        super.onCreate(savedInstanceState)
        val updateScheduler = Skylight.applicationComponent.getUpdateScheduler()
		val notificationsKey = resources.getString(R.string.pref_notifications_key)
		notificationsChangedListener = NotificationsChangedListener(updateScheduler, notificationsKey)
        preferenceManager.sharedPreferences.registerOnSharedPreferenceChangeListener(notificationsChangedListener)
        addPreferencesFromResource(R.xml.preferences)
        // TODO Daggerify
    }

    override fun onDestroy() {
        Log.v(TAG, "onDestroy")
        preferenceManager.sharedPreferences.unregisterOnSharedPreferenceChangeListener(notificationsChangedListener)
        super.onDestroy()
    }

    companion object {
        private val TAG = SettingsFragment::class.java.simpleName
    }
}

private class NotificationsChangedListener(
		private val updateScheduler: UpdateScheduler,
		private val notificationsKey: String
) : SharedPreferences.OnSharedPreferenceChangeListener {

	override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
		if (key == this.notificationsKey) {
			updateScheduler.setupBackgroundUpdates()
		}
	}
}
